cookbook/security/user_checkers.rst

```diff
@@ -0,0 +1,221 @@
+:class:`UserCheckerInterface Symfony\\Component\\Security\\Core\\UserCheckerInterface`,
+this interface is required. If one or more conditions are not met, an exception should
+be throw which extends the
+:class:`AccountStatusException Symfony\\Component\\Security\\Core\\Exception\\AccountStatusException`
+
+.. code-block:: php
+
+    namespace App\Security\User;
+
+    use Symfony\Component\Security\Core\User\UserCheckInterface;
+
+    class AppUserChecker UserCheckerInterface
+    {
+        public function checkPreAuth(UserInterface $user)
+        {
+            if (!$user instanceof AppUser) {
+                return;
+            }
+
+            // user is deleted, show a generic Account Not Found message.
+            if ($user->isDeleted()) {
+                throw new AccountDeletedException('...');
+            }
+        }
+
+        public function checkPostAuth(UserInterface $user)
+        {
+            if (!$user instanceof AppUser) {
+                return;
+            }
+
+            // user account is expired, the user may be notified
+            if ($user->isExpired()) {
+                throw new AccountExpiredException('...');
+            }
+        }
+    }
+
+Enabling the Custom User Checker
+--------------------------------
+
+All that's left to be done is creating a service definition and configuring
+this in the firewall configuration. Configuring the service is done like any
+other service:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/services.yml
+        services:
+            app.user_checker:
+                class: App\Security\User\AppUserChecker
+
+    .. code-block:: xml
+
+        <!-- app/config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <service id="app.user_checker" class="App\Security\User\AppUserChecker">
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // app/config/services.php
+        use Symfony\Component\DependencyInjection\Definition;
+
+        $userChecker = new Definition('App\Security\User\AppUserChecker');
+        $container->setDefinition('app.user_checker', $userChecker);
+
+All that's left to do is add the checker to the desired firewall where the value
+is the service id of your user checker:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/security.yml
+
+        # ...
+        security:
+            firewalls:
+                secured_area:
+                    pattern: ^/
+                    user_checkers:
+                        - "app.user_checker"
```

Comment:
I prefer the short array notation, but I don't know which is the practice for Symfony documentation:

```yaml
user_checkers: ['app.user_checker']
```